sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -39,137 +64,178 @@ public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, ICompar
-        }
-#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+            if (resourceId.Length == 1 && resourceId[0] == Separator)
+            {
+                //this is the same as Root but can't return that since this is a ctor
+                Init(null, ResourceType.Tenant, string.Empty, false, null);
+            }
+            else
+            {
+                ReadOnlySpan<char> remaining = resourceId.AsSpan();
 
-        private ResourceIdentifier(ResourceIdentifier parent, string resourceTypeName, string resourceName)
-            : this(parent, ChooseResourceType(resourceTypeName, parent), resourceName, false)
-        {
+                if (!remaining.StartsWith(SubscriptionStart.AsSpan()) && !remaining.StartsWith(ProviderStart.AsSpan()))
+                    throw new ArgumentOutOfRangeException(nameof(resourceId), $"The ResourceIdentifier must start with {SubscriptionStart} or {ProviderStart}.");
+
+                remaining = remaining.Slice(1);
+                //trim trailing '/' off the end if it exists
+                if (remaining[remaining.Length - 1] == Separator)
+                    remaining = remaining.Slice(0, remaining.Length - 1);
+
+                ReadOnlySpan<char> nextWord = PopNextWord(ref remaining);
+
+                //we know we at least have 1 ResourceIdentifier in the tree here
+                ResourceIdentifierParts nextParts = GetNextParts(Root, ref remaining, ref nextWord);
+                while (!nextWord.IsEmpty)
+                {
+                    //continue to get the next ResourceIdentifier in the tree until we reach the end which will be 'this'
+                    ResourceIdentifier newParent = new ResourceIdentifier(nextParts.Parent, nextParts.ResourceType, nextParts.ResourceName, nextParts.IsProviderResource, nextParts.SpecialType);
+                    nextParts = GetNextParts(newParent, ref remaining, ref nextWord);
+                }
+
+                //initialize ourselves last
+                Init(nextParts.Parent, nextParts.ResourceType, nextParts.ResourceName, nextParts.IsProviderResource, nextParts.SpecialType);
+                _stringValue = resourceId;
+            }
         }
 
-        private ResourceIdentifier(ResourceIdentifier parent, string providerNamespace, string resourceTypeName, string resourceName)
-            : this(parent, new ResourceType(providerNamespace, resourceTypeName), resourceName, true)
+        private ResourceIdentifier(ResourceIdentifier? parent, ResourceType resourceType, string resourceName, bool isProviderResource, SpecialType? specialType)
         {
+            Init(parent, resourceType, resourceName, isProviderResource, specialType);
         }
+#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
```

Comment:
Which type is this about do you know?

Reply:
@annelo-msft  originally suggested this so we didn't duplicate the property initialization in both ctors.  She has changed her mind and asked me to revert it which I did here along with the nullable parent https://github.com/Azure/azure-sdk-for-net/pull/26237

Reply:
Ok.